File: Data Structures & Algorithms/reorder-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

#Brainstorm
# Time O(n)
# Space O(1)

# Plan
# 1. find middle of list
# 2. reverse list
# 3. merge the lists
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        default = ListNode(0, head)
        slow = head
        fast = head
        # 1. find middle of list
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        # 2. reverse list
        cur = slow.next
        prev = None
        slow.next = None
        while cur:
            next = cur.next
            cur.next = prev
            prev = cur
            cur = next

        # 3. merge the lists
        first = head
        second = prev

        while second:
            # Save the next nodes before we change pointers
            first_next = first.next
            second_next = second.next
            
            # Insert second node after first node
            first.next = second
            second.next = first_next
            
            # Move to the next pair of nodes to merge
            first = first_next
            second = second_next

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.swift
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     var val: Int
 *     var next: ListNode?
 *     init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func mergeTwoLists(_ list1: ListNode?, _ list2: ListNode?) -> ListNode? {
        if list1 == nil {
            return list2
        }
        if list2 == nil {
            return list1
        }

        var l1 = list1
        var l2 = list2

        var res: ListNode? = ListNode(0)
        let returnedRes = res

        while l1 != nil && l2 != nil {
            if l1!.val <= l2!.val {
                res!.next = ListNode(l1!.val)
                l1 = l1!.next
            } else {
                res!.next = ListNode(l2!.val)
                l2 = l2!.next
            }
            res = res!.next
        }

        while l1 != nil {
            res!.next = ListNode(l1!.val)
            l1 = l1!.next
            res = res!.next
        }

        while l2 != nil {
            res!.next = ListNode(l2!.val)
            l2 = l2!.next
            res = res!.next
        }

        return returnedRes!.next
    }
}
